pep-0484.txt

```diff
@@ -276,6 +276,33 @@ implemented by deferring to ``isinstance(x, collections.abc.Callable)``.
 However, ``isinstance(x, typing.Callable[...])`` is not supported.
 
 
+Decorators
+----------
+
+Decorators can modify the types of the functions or classes they
+decorate. To annotate a decorated declaration with the eventual
+decorated type, put a type comment on the line with the decorator::
+
+  from typing import ContextManager, Generator
+  from contextlib import contextmanager
+
+  class DatabaseSession: ...
+
+  @contextmanager  # type: Callable[[str], ContextManager[DatabaseSession]]
```

Comment:
I don't like the overloading of cast -- it feels too magical. However, I might prefer using a new decorator instead of a type comment, primarily because Python 3.6 mostly doesn't require type comments any more and they feel like a legacy feature, and also because a decorator would allow splitting the annotation to multiple lines. 

I wouldn't call this a cast but a declaration. A type checker should still check that the inferred signature is compatible with the declared one. This is unlike casts, which a type checker should blindly accept.

Here are ideas for the syntax:
```py
@decorator_type(Callable[[int], str])
@mydecorator
def f(...): ...

@decorated_type(Callable[[int], str])
@mydecorator
def f(...): ...

@decorated_type([int], str)    # convenience 2-argument syntax that doesn't require Callable[...]
@mydecorator
def f(...): ...

@decorated_signature([int], str)
@mydecorator
def f(...): ...
```